hw/ip/otbn/util/shared/control_flow.py

```diff
@@ -153,27 +164,14 @@ def get_edges(self, pc: int) -> List[ControlLoc]:
         '''
         return self.get_entry(pc)[1]
 
-    def get_cycles(self,
-                   pc: int,
-                   path: List[int] = []) -> Dict[int, List[int]]:
-        '''Returns any (non LOOP/LOOPI) control-flow cycles in this graph.'''
-        cycles: Dict[int, List[int]] = {}
-        sec, edges = self.get_entry(pc)
-        for loc in edges:
-            if isinstance(loc, LoopStart):
-                pc = loc.loop_start_pc
-            elif loc.is_special():
-                # Skip Ret, Ecall, ImemEnd, LoopEnd
-                continue
-            else:
-                pc = loc.pc
-            if pc in path:
-                # Found a cycle; report it
-                cycles.setdefault(pc, []).append(sec.end)
-            else:
-                pcs = list(range(sec.start, sec.end + 4, 4))
-                cycles.update(self.get_cycles(pc, path + pcs))
-        return cycles
+    def get_cycle_starts(self) -> Set[int]:
+        '''Returns start PCs of all marked cycles in the graph.'''
+        out = set()
+        for pc, entry in self.graph.items():
+            for edge in entry[1]:
+                if isinstance(edge, Cycle):
+                    out.add(edge.pc)
```

Comment:
Just wondering: why can't you use `entry[0]` here? (which would also avoid duplicates for free, without having to mess around with sets)

Reply:
Here, `entry[0]` is a `CodeSection` that represents the straightline block starting at `pc`. It ends in some kind of control-flow instruction, and `entry[1]` is a list of the next locations that could be reached from that instruction. So I don't have the information about `entry[0]` that I would need -- namely, what type of "control-flow location" it is. I'm not sure off the top of my head if it's true that I could restructure this to have that information (e.g. maybe adding a third item to the `entry` tuple), but in the current structure that information isn't possible to determine without looking at the edge entry.

Reply:
Ahhh! Now I understand. Thanks for the explanation.